### PR TITLE
Enable SyncedList to handle general objects

### PR DESCRIPTION
**Pull request by b-butler**
## Description

This enables `SyncedList` to handle objects that do not have a concept of attaching or detaching.

## Motivation and context
This change was inspired by walls and the need to manage pure Python wall specifications.

## How has this been tested?

Existing tests are modified to fit the new `SyncedList` implementation and pass.

## Change log

None needed (hopefully no one was depending on internal implementations of `SyncedList` already).

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by b-butler** [comment]
@joaander and @pepak13 This PR is necessary for #1020, so let me know if you can't review it or need some help.

**Review by pepak13 — APPROVED** [review]
Looks good to me

**Comment by bdice** [comment]
@b-butler I will review on/after Wednesday.

**Review by joaander — APPROVED** [review]
This looks fine to me. It is all internal changes to the helper methods to sync data structures between C++ and Python. @bdice is this something you think you need to review in detail before merging?

**Comment by bdice** [comment]
@joaander Taking a look now! I am interested in seeing the implementation.

**Review by bdice — APPROVED** [review]
I'm unsure about the large amount of logic for indexing in `_handle_int`, `_handle_slice`, `_islice`, and `_islice_index`. I think that all of this can be handled with judicious use of `range` and `itertools.islice`? There might be a way to reduce the number of such functions (and reduce the number of edge cases they handle) by 25%-50%. Since this has two approvals, I won't hold it up, but it could be a worthy refactor.

File: hoomd/util.py
Comment (b-butler) [reply hoomd/util.py:35]: To allow for both slice object and slice specifications, this let's a slice object be passed as well as the start, stop, and step. You could use keyword arguments like  `slice`, or `start`, `stop`, `step`. I just thought that positional arguments here were easier.

**Comment by b-butler** [comment]
@bdice we could use `islice` if we constructed an "index" list every time we needed the indices. That would reduce the logic. The main reason, I implemented the functions the way I did was to A) natively handle slice objects which isn't that difficult if you just explicit call for its attributes in `islice` or to return indices rather than the items from the sequence itself which as mentioned could be handled with an index list and `islice`. I believe at the time I made the changes, I didn't do that to avoid the extra list construction, but that is probably an acceptable trade-off for removing the logic from hoomd.

File: hoomd/util.py
Comment (b-butler) [reply hoomd/util.py:35]: Also, `itertools.islice` uses positional arguments.

Comment (bdice) [reply hoomd/util.py:36]: But `itertools.islice` doesn't accept both slice objects and start/stop/step. 😉 This isn't a big deal. I just didn't understand the reasoning. This PR is good enough to merge if it's holding up other work, but there is some excess code that may someday be technical debt.

File: hoomd/util.py
Comment (bdice) [inline hoomd/util.py:45]: The `list` call might be unnecessary, based on my quick tests. `range` objects can be sliced directly to produce another `range` object.

However, you must ensure that all values are positive here, you can't allow negative indexing in `islice` like `stop=-1`.

Separately, do you want to handle the case of a single positional argument `stop`? That doesn't seem to be handled (or documented) right now.

I think it might need to look like this?
```suggestion
    if len(args) == 1:
        args[0] = _handle_index(args[0])
    else:
        args = [_handle_index(args[0]), _handle_index(args[1]), args[2]]
    yield from itertools.islice(list(range(len(sequence))), *args)
```

Comment (b-butler) [reply hoomd/util.py:45]: Hmm, I didn't realize that islice didn't handle the full slice specification... I might just revert the change then, because that means, I will have to effectively code the full logic anyways to handle all slices. Also, `stop` is handled since `islice(seq, stop)` works and we just pass the args along unless the passed object was a slice object.

File: hoomd/data/syncedlist.py
Comment (b-butler) [reply hoomd/data/syncedlist.py:223]: Good catch.

File: hoomd/data/syncedlist.py
Comment (b-butler) [reply hoomd/data/syncedlist.py:154]: That does feel somewhat "wrong", but thanks for the suggestion.

**Comment by b-butler** [comment]
@bdice, thanks for the help. I didn't know that range objects could be indexed by slices. This greatly simplifies the code.

**Comment by joaander** [comment]
@b-butler Feel free to merge when you've made all modifications you plan to make.

**Pull request merged**
